npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,428 @@
+    - domains:
+      - "my-service.com"
+      - "*.cloud-provider.io"
+    ports:
+    - portNumber:
+        protocol: TCP
+        port: 443
+```
+
+#### Maintaining an allowlist of domains
+
+There are a couple ways to maintain an allowlist:
+
+This example, includes the DENY rule in the same ANP object. It's also possible
+to use another ANP object with a lower priority (e.g. `100` in this example):
+```yaml
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: allow-my-service-egress
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchLabels:
+        kubernetes.io/metadata.name: "monitoring"
+  egress:
+  - name: "allow-to-my-service"
+    action: "Allow"
+    to:
+    - domains:
+      - "my-service.com"
+      - "*.cloud-provider.io"
+    ports:
+    - portNumber:
+        protocol: TCP
+        port: 443
+  - name: "default-deny"
+    action: "Deny"
+    to:
+    - networks:
+      - "0.0.0.0/0"
+```
+
+This example uses a default-deny BaselineAdminNetworkPolicy to create the
+allowlist:
+```yaml
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: allow-my-service-egress
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchLabels:
+        kubernetes.io/metadata.name: "monitoring"
+  egress:
+  - name: "allow-to-my-service"
+    action: "Allow"
+    to:
+    - domains:
+      - "my-service.com"
+      - "*.cloud-provider.io"
+    ports:
+    - portNumber:
+        protocol: TCP
+        port: 443
+---
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: BaselineAdminNetworkPolicy
+metadata:
+  name: default
+spec:
+  subject:
+    namespaces: {}
+  ingress:
+    - action: Deny
+      to:
+      - networks:
+        - "0.0.0.0/0"
+```
+
+### Expected Behavior
+
+1. A FQDN egress policy does not grant the workload permission to communicate
+   with any in-cluster DNS services (like `kube-dns`). A separate rule needs to
+   be configured to allow traffic to any DNS servers.
+1. FQDN policies should not affect the ability of workloads to resolve domains,
+   only their ability to communicate with the IP backing them.
+   *  For example, if a policy allows traffic to `kubernetes.io`, any selected
+      Pods can still resolve `wikipedia.org` or
+      `my-services.default.svc.cluster.local`, but can not send traffic to them
+      unless allowed by a different rule.
+1. Each implementation will provide guidance on which DNS name-server is
+   considered authoritative for resolving domain names. This could be the
+   `kube-dns` Service or potentially some other DNS provider specified in the
+   implementation's configuration.
+1. DNS record querying and lifetimes:
+   *  Pods are expected to make a DNS query for a domain before sending traffic
+      to it. If the Pod fails to send a DNS request and instead just sends
+      traffic to the IP (either because of caching or a static config), traffic
+      is not guaranteed to flow.
```

Comment:
is this implementable? this means the implementation has to cache all possible IPs of all the possible domains, no?